### PR TITLE
Change styling of slider widgets

### PR DESCRIPTION
**Pull request by ianhi**
## PR Summary
Closes: https://github.com/matplotlib/matplotlib/issues/19256

Updates the styling of the slider widgets to look like this:
Peek 2021-01-12 17-59

1. Made the `poly` patch smaller
2. Added a circle handle to indicate that the sliders can be grabbed with the mouse
   - ~will change color when grabbed~
3. removed the black border of the `Axes` and indicate the length of the slider with rectangle patch. 
    - This is skinnier than previously.

## PR Checklist

- [NA?] Has pytest style unit tests (and `pytest` passes).
- [X] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [I guess?] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [X] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by QuLogic** [comment]
Is `_above_rect` for the grey part? Why not set a default `Axes` facecolor instead?

**Comment by ianhi** [comment]
I did it that way for two reasons:
1. If you make the blue polygon have an `alpha != 1` then setting the facecolor of the axis looks bad.
2. I think it looks better when the line the slider is on is smaller and this way the existing slider axis size from the example (and thus probably most peoples code) will not need to be modified for that visual change 

I suppose those probably don't fully justify the added complexity though

**Comment by ianhi** [comment]
Just tried out with using the `facecolor` approach. I found that it doesn't look good with the default scaling (first slider), but that if I made the `Axes` used to make the slider smaller in order to look nicer it was significantly more difficult to click on due to the axis being so small. That's compounded by the fact that clicking on the handle when it clips outside doesn't register as moving the slider unfortunately as the events are all tied to the axis.

image

**Comment by ianhi** [comment]
I guess it would make sense to simplify by just having a single static background rect

**Comment by WeatherGod** [comment]
can we have a bigger difference in brightness for the right and left
rectangles? They are too close to easily distinguish.

**Comment by ianhi** [comment]
> can we have a bigger difference in brightness for the right and left
rectangles? They are too close to easily distinguish.

Fine by me. Any specific suggestions for colors? The left rectangle is currently just whatever is first in the users color cycle as it is generated by `ax.axvspan`.

**Comment by ianhi** [comment]
How about `lightgrey`? The bottom slider here:
image

File: lib/matplotlib/widgets.py [hunk 1]
```diff
@@ -404,11 +424,31 @@ def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt=None,
         self.val = valinit
         self.valinit = valinit
         if orientation == 'vertical':
-            self.poly = ax.axhspan(valmin, valinit, 0, 1, **kwargs)
-            self.hline = ax.axhline(valinit, 0, 1, color=initcolor, lw=1)
+            self.bkd = Rectangle((.25, 0), .5, 1, transform=ax.transAxes, facecolor=bkd_color)
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:427]: Can we have a more telling name?
```suggestion
            self.track = Rectangle((.25, 0), .5, 1, transform=ax.transAxes, facecolor=bkd_color)
```
seems to be the technical term.

File: lib/matplotlib/widgets.py [hunk 2]
```diff
@@ -472,7 +512,13 @@ def _update(self, event):
                event.inaxes != self.ax)):
             self.drag_active = False
             event.canvas.release_mouse(self.ax)
+            self._handle.set_markeredgecolor(self._handle_edgecolor)
+            self._handle.set_markerfacecolor(self._handle_color_released)
+            self.ax.figure.canvas.draw_idle()
             return
+        if event.name == 'button_press_event':
+            self._handle.set_markeredgecolor(self._handle_color_grabbed)
+            self._handle.set_markerfacecolor(self._handle_color_grabbed)
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:521]: I'm not sure if this action helpful. If sliders have GUI feedback, it's usually already done on hovering to indicate that you have an control element under the cursor. IMHO changing the color after clicking is too late and serves no purpose.

For simplicity, I'd refrain from changing the color at all.

**Comment by WeatherGod** [comment]
I like the lightgrey. Works very well.

Another thing to be aware of is that while white (or rather transparent on
top of the standard gui window color, which is white) is the default, there
are people who change these defaults or are embedding matplotlib in other
apps. I haven't looked at the changes yet. but how easy is it going to be
to tweak these colors?

File: lib/matplotlib/widgets.py (hunk 2 above)
Comment (ianhi) [reply lib/matplotlib/widgets.py:521]: I did this because I was copying the behavior of ipywidgets sliders without much critical thought :grimacing:.

I have a slight preference for them changing color after selection vs not but I'll happily remove it when I next push here. Reducing the number of new configuration kwargs seems like a good thing.

Comment (ianhi) [reply lib/matplotlib/widgets.py:521]: That said changing as you mouse over the axes could be nice. I imagine this could be accomplished by using `axes_enter_event` and `axes_leave_event`

File: lib/matplotlib/widgets.py (hunk 1 above)
Comment (ianhi) [reply lib/matplotlib/widgets.py:427]: Ah that's a much better name thanks. I was struggling to find the correct term.

**Comment by ianhi** [comment]
> I haven't looked at the changes yet. but how easy is it going to be
to tweak these colors?

The track color has a `kwarg` dedicated to it + the underlying rect is accesible as a public attribute of the slider, and the blue color has always configurable as `**kwargs` has always been passed on to `axvspan`.

File: lib/matplotlib/widgets.py [hunk 3]
```diff
@@ -404,11 +424,31 @@ def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt=None,
...
+            handleXY = [[0.5], [valinit]]
         else:
-            self.poly = ax.axvspan(valmin, valinit, 0, 1, **kwargs)
-            self.vline = ax.axvline(valinit, 0, 1, color=initcolor, lw=1)
+            self.bkd = Rectangle((0, .25), 1, .5, transform=ax.transAxes, facecolor=bkd_color)
+            ax.add_patch(self.bkd)
+            self.poly = ax.axvspan(valmin, valinit, .25, .75, **kwargs)
+            self.vline = ax.axvline(valinit, .25, .75, color=initcolor, lw=1)
+            handleXY = [[valinit], [0.5]]
+        self._handle, = ax.plot(
```
Comment (ianhi) [inline lib/matplotlib/widgets.py:438]: Should the handle be a public attribute in order to enable full customizability?

File: lib/matplotlib/widgets.py (hunk 2 above)
Comment (timhoffm) [reply lib/matplotlib/widgets.py:521]: If you want an effect, see also `Button` for a mouse-over effect. Though here you might want to additionally check if the cursor is on the knob.

We could also say that the "active" color is the same color as the bar (C0). That would still save us the extra argument. One can always add more configurability later when the need arises.

File: lib/matplotlib/widgets.py (hunk 3 above)
Comment (timhoffm) [reply lib/matplotlib/widgets.py:438]: For the moment, I'd keep it private. It's a bit arcane because it's a marker represented by a one-element Line2D. This would need extra explanation. I'm also not sure if the handle should or should not scale with the slider when the figure is rescaled.

Comment (ianhi) [reply lib/matplotlib/widgets.py:438]: Thats what I thought, but it is a break from the other visual elements of the slider which are public.

> I'm also not sure if the handle should or should not scale with the slider when the figure is rescaled.

I'm starting to think more and more that sliders should be specified in inches rather than whatever they are now (I guess technically in figure coords?). Though having part of the slider scale while the other doesn't seems like a bad inconsistency, so they probably ought to scale. Not sure how best to achieve marker scaling thought. (I couldn't figure out how to make a circle with consistent aspect ratio without resorting to specifying it in inches)

File: lib/matplotlib/widgets.py (hunk 2 above)
Comment (ianhi) [reply lib/matplotlib/widgets.py:521]: > ipywidgets use color to indicate focus (e.g. which control reacts to the keyboard). I don't think our widgets have a concept of focus.


aaah. I'm fully convinced by this, I'll remove it as it's bad to indicate something that is isn't there.

**Comment by ianhi** [comment]
Also: heads up to @t-makaro you may be interested in this on account of animatplot.

File: lib/matplotlib/widgets.py (hunk 3 above)
Comment (timhoffm) [reply lib/matplotlib/widgets.py:441]: > Thats what I thought, but it is a break from the other visual elements of the slider which are public.

I'd be pragmatic and wait until people ask for configurability; in particular also because we might need to change the object representing the knob if we do something about scaling (which doesn't have to be right now in this PR).

#### Size specification

> I'm starting to think more and more that sliders should be specified in inches rather than whatever they are now (I guess technically in figure coords?).

*tl;dr*: Leave as is for now.

Yes, it's figure coords. The advantage of this is that you can line up the sliders with any Axes, because they also use figure coords. IMHO inches are not a great idea, because the rendering position then depends on `dpi`. Inches is rather an abstract quantity when you render to the screen as you do with widgets (I know our figure size is in inches, but unless people are explicitly changing it, they may not even know.

#### Scaling

*tl;dr*: Leave as is for now.

- Sliders have to scale their long axis. Otherwise the layout will most certainly be broken when the figure is scaled.
- Maybe they would not necessarily have to scale their track width. But it's easier to implement scaling the track width. And it's not that bad.
- We don't have an object that stays circular under arbitrary figure scalings but also scales. Actually you have to put additional assumptions in to uniquely specify it: If the aspect ratio changes under scaling, you have two different scaling factors for x and y. How do you reduce them to one scaling factor of the circle?
One would have to write a specialized Artist for this. IMHO that's not worth it (and is anyway beyond the scope of this PR). An alternative would be using a marker and adapting it's size based on resize events (but to be checked how that interacts with draw).

  For practical reasons, stay with the circular marker. Unless people are massively scaling (> 2x) this still looks good.

File: lib/matplotlib/widgets.py [hunk 4]
```diff
@@ -380,6 +382,19 @@ def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt=None,
...
+            The color of the backgrond track. The track is accesible for
+            further styling via the *track* attribute.
+
+        handle_facecolor : color, default: 'white'
+            The facecolor of the circular slider handle.
+
+        handle_edgecolor : color, default: '.75'
+            The edgecolor of the circle slider handle.
+
+        handle_size : int, deafult: 10
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:395]:
```suggestion
        handle_size : int, default: 10
```

File: lib/matplotlib/widgets.py [hunk 5]
```diff
@@ -380,6 +382,19 @@ def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt=None,
...
+            The size of circular slider handle.
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:396]:
```suggestion
            The size of the circular slider handle in points.
```

File: lib/matplotlib/widgets.py [hunk 6]
```diff
@@ -598,6 +641,23 @@ def __init__(
         orientation : {'horizontal', 'vertical'}, default: 'horizontal'
             The orientation of the slider.
 
+        track_color : color, default: 'lightgrey'
+            The color of the backgrond track. The track is accesible for
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:645]:
```suggestion
            The color of the background track. The track is accessible for
```

File: lib/matplotlib/widgets.py [hunk 7]
```diff
@@ -598,6 +641,23 @@ def __init__(
...
+            further styling via the *track* attribute.
+
+        handle_color_grabbed : color or None, default: None
+            The color of the circular slider handles when grabbed. If *None*
+            the same color as active part of the slider will be used.
+
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:651]: Parameter has been removed.

File: lib/matplotlib/widgets.py [hunk 8]
```diff
@@ -598,6 +641,23 @@ def __init__(
...
+        handle_facecolor : color, default: 'white'
+            The facecolor of the circular slider handle.
+
+        handle_edgecolor : color, default: '.75'
+            The edgecolor of the circular slider handles.
+
+        handle_size : int, deafult: 10
+            The size of circular slider handles.
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:659]:
```suggestion
            The size of the circular slider handles in points.
```

File: lib/matplotlib/widgets.py [hunk 9]
```diff
@@ -380,11 +382,24 @@ def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt=None,
...
+
+        handle_size : int, default: 10
+            The size of the circular slider handle in points.
+
         Notes
         -----
         Additional kwargs are passed on to ``self.poly`` which is the
-        `~matplotlib.patches.Rectangle` that draws the slider knob.  See the
-        `.Rectangle` documentation for valid property names (``facecolor``,
+        `~matplotlib.patches.Polygon` that draws the slider knob.  See the
```
Comment (ianhi) [inline lib/matplotlib/widgets.py:401]: https://matplotlib.org/devdocs/api/_as_gen/matplotlib.pyplot.axvspan.html

`ax{h,v}span` returns a `Polygon` not a `Rectangle` patch.

File: lib/matplotlib/widgets.py (hunk 3 above)
Comment (ianhi) [reply lib/matplotlib/widgets.py:441]: Thanks for the extensive explanations!

**Comment by ianhi** [comment]
rebased + squashed and updated the GIF on the first post.
I also added a change the notes of the slider docstring as the `self.poly` is a `Polygon` not a `Rectangle`

**Comment by ianhi** [comment]
I've also reworked the `slider_snap_demo` to be an example of many sliders that demonstrates how to style them. (see it here: https://github.com/ianhi/matplotlib/blob/slider-styling-example/examples/widgets/slider_styling.py)
image


Should I include that here or add that as a followup PR?

File: lib/matplotlib/widgets.py [hunk 10]
```diff
@@ -404,11 +419,33 @@ def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt=None,
...
-            self.poly = ax.axhspan(valmin, valinit, 0, 1, **kwargs)
-            self.hline = ax.axhline(valinit, 0, 1, color=initcolor, lw=1)
+            self.track = Rectangle(
+                (.25, 0), .5, 1,
+                transform=ax.transAxes,
+                facecolor=track_color
+            )
+            ax.add_patch(self.track)
+            self.poly = ax.axhspan(valmin, valinit, .25, .75, **kwargs)
+            self.hline = ax.axhline(valinit, .25, .75, color=initcolor, lw=1)
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:429]:
```suggestion
            self.hline = ax.axhline(valinit, .2, .8, color=initcolor, lw=1)
```
The init line extends one pixel more to the bottom (or right for vertical sliders), which looks a bit awkward.

I suspect that the effect comes from different to-pixel mappings of lines and rectangles (lines render including both end points; i.e. two lines ending in the same point would draw on top of each other. But rectangles are designed to not overlap if they are side-by-side). - This is just speculation, but I've checked and excluded some other possible reasons (capstyle and snapping).

The cheap solution to get away with a reasonably looking init marker is to let it extend slightly on both sides. The pixel difference is still there, but 2 vs. 3 is less noticeable than 0 vs. 1.

File: lib/matplotlib/widgets.py [hunk 11]
```diff
@@ -329,7 +329,9 @@ class Slider(SliderBase):
     def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt=None,
                  closedmin=True, closedmax=True, slidermin=None,
                  slidermax=None, dragging=True, valstep=None,
-                 orientation='horizontal', *, initcolor='r', **kwargs):
+                 orientation='horizontal', *, initcolor='r',
+                 track_color='lightgrey', handle_facecolor='white',
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:333]: I'm contemplating whether we should collect all handle attributes into one parameter, in the form
`handle={'color': 'lightgrey', facecolor='white', edgecolor='.75', size=10}`. That's a bit more indirect and less discoverable (we need to write down the supported keys in the docstring).

But it would be easier to expand on this later. E.g. we could use `handle='none'`, or `handle={'type': 'rect'}`.

The decision depends a bit on whether we assume that further options will be added.

**Comment by timhoffm** [comment]
> Should I include that here or add that as a followup PR?

I recommend a separate PR, because that makes reviewing simpler. (The longer a PR is, the less likely somebody takes the time to review. And the relation is non-linear. :stuck_out_tongue_closed_eyes: )

File: lib/matplotlib/widgets.py (hunk 11 above)
Comment (ianhi) [reply lib/matplotlib/widgets.py:333]: In this case the default argument would become `None`?

I'm a fan of adding flexibility here, especially because the handle is a private attribute and predicting the future is well known to be a fraught activity. As for it being indirect, that is indeed unfortunate, but I suspect that the proposed slider styling example will go a long way towards fixing that.

Are the other examples of an argument like this that I can use for inspiration for writing the docstring?  (Maybe there should be a short blurb somewhere in the contributing docs explaining how to approach this and other non-trivial docstrings?)

File: lib/matplotlib/widgets.py (hunk 10 above)
Comment (ianhi) [reply lib/matplotlib/widgets.py:429]: I agree that this simple solution is the way to go. Though I actually feel it is more noticeable in the horizontal slider in your second screenshot. Maybe this is due to your screenshot program but it looks as though on the vertical slider it is 2 vs 3 (or even the same):
image

whereas on the horizontal slider:
image

File: lib/matplotlib/widgets.py (hunk 11 above)
Comment (timhoffm) [reply lib/matplotlib/widgets.py:333]: The examples we have are unfortunately not very well documented:

- the parameter `fontdict` e.g. in [set_title](https://matplotlib.org/devdocs/api/_as_gen/matplotlib.axes.Axes.set_title.html)
- the `bbox` property of [Text](https://matplotlib.org/devdocs/api/text_api.html#matplotlib.text.Text)
- the `arrowprops` in [annotate](https://matplotlib.org/devdocs/api/_as_gen/matplotlib.axes.Axes.annotate.html)

I'd do it like this:

~~~
        handle : dict
            Properties of the slider handle. Supported values are

            ========= ===== ======= ========================================
            Key       Value Default Description
            ========= ===== ======= ========================================
            facecolor color 'white' The facecolor of the slider handle.
            edgecolor color '.75'   The edgecolor of the slider handle.
            size      int   10      The size of the slider handle in points.
            ========= ===== ======= ========================================

~~~
I hope it's valid ReST - didn't check - but you get the point.

Comment (ianhi) [reply lib/matplotlib/widgets.py:333]: Got it thanks!

> Supported values are

Should we be limiting this? I was imagining that this would be passed as `ax.plot(..., **handle)` that way if anyone feels very passionate about restyling the handle they can do whatever they want.

Comment (timhoffm) [reply lib/matplotlib/widgets.py:333]: You need to `marker_props = {f'marker{key}': val for key, val in handle}` to be able to `ax.plot(..., **marker_props)`, which effectively is a strong limit. If you want a better error message, you'd need explicit checks. In that case, you should write a function `check_dict_keys()` similar to the existing `check_*` methods.

Comment (ianhi) [reply lib/matplotlib/widgets.py:333]: I'm not sure I fully understood your last comment, but I decided to go with your initial suggestion. I think it makes the docstring far clearer to the user.

Comment (timhoffm) [reply lib/matplotlib/widgets.py]: What I was trying to say: You are converting the keys in handle to 'marker{key}' and pass that to `plot()`. Thus, only keys for which a 'marker{key}' is a valid parameter to `plot()` will be accepted. This already limits the possible handle parameters to more or less reasonable values.

If you passe `handle={'foo': 1}`, you get a message 
> 'Line2D' object has no property 'markerfoo'

which is fair but not overly precise. In case you want a better error message such as "'foo' is not a valid handle parameter", you'd need to validate the handle explicitly.

File: lib/matplotlib/widgets.py [hunk 12]
```diff
@@ -403,12 +419,41 @@ def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt=None,
...
         self.val = valinit
         self.valinit = valinit
+
+        marker_props = {}
+        defaults = {'facecolor':'white', 'edgecolor':'.75', 'size':10}
+        if handle_style is not None:
+            for k in ['facecolor', 'edgecolor', 'size']:
+                marker_props[f'marker{k}'] = handle_style.get(k, defaults[k])
+        else:
+            marker_props = {f'marker{k}': v for k, v in defaults.items()}
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:429]: You can simplify to:

```suggestion
        defaults = {'facecolor':'white', 'edgecolor':'.75', 'size':10}
        marker_props = {f'marker{k}': v for k, v in {**defaults, **handle_style}}
```
This has one desired functional difference: Invalid keys in `handle_style` will be promoted to `marker{key}` and thus raise when passed to plot. We don't want to accept arbitrary keys for `handle_style` and silently discard them.

File: lib/matplotlib/widgets.py (hunk 11 above)
Comment (ianhi) [reply lib/matplotlib/widgets.py]: Thank you for explaining! I've tried include this in the docstrings.

File: lib/matplotlib/widgets.py [hunk 13]
```diff
@@ -380,11 +381,29 @@ def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt=None,
...
+            Properties of the slider handle. Default values are
+
+            ========= ===== ======= ========================================
+            Key       Value Default Description
+            ========= ===== ======= ========================================
+            facecolor color 'white' The facecolor of the slider handle.
+            edgecolor color '.75'   The edgecolor of the slider handle.
+            size      int   10      The size of the slider handle in points.
+            ========= ===== ======= ========================================
+            Other values will be transformed as marker{foo} and passed to the
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:398]:
```suggestion

            Other values will be transformed as marker{foo} and passed to the
```
ReST requires a blank line after a table.

File: lib/matplotlib/widgets.py [hunk 14]
```diff
@@ -598,11 +648,29 @@ def __init__(
...
+            Properties of the slider handles. Default values are
+
+            ========= ===== ======= =========================================
+            Key       Value Default Description
+            ========= ===== ======= =========================================
+            facecolor color 'white' The facecolor of the slider handles.
+            edgecolor color '.75'   The edgecolor of the slider handles.
+            size      int   10      The size of the slider handles in points.
+            ========= ===== ======= =========================================
+            Other values will be transformed as marker{foo} and passed to the
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:665]:
```suggestion

            Other values will be transformed as marker{foo} and passed to the
```

**Review by timhoffm — COMMENTED** [review]
There are more warnings causing the doc build to fail due to #19415.

Please rebase so that that fix is picked up.

File: lib/matplotlib/widgets.py (hunk 13 above)
Comment (ianhi) [reply lib/matplotlib/widgets.py:399]: Thanks! I applied these in the rebase

**Comment by ianhi** [comment]
> Please rebase so that that fix is picked up.

rebased - thanks for the heads up

**Comment by ianhi** [comment]
Hi,

I think I've addressed all of the requested revisions. Is it ok to remove the `needs revision` tag?

**Comment by anntzer** [comment]
> Is it ok to remove the needs revision tag?

Yes.  (Basically whenever you consider you handled the revision requests.)

**Comment by timhoffm** [comment]
> Well, perhaps with the exception of the issue of the red line offset #19265 (comment) though I don't think that that is waiting on me

Actually, I was waiting on that. IMHO uneven extensions are at least better than one side ending exactly on the line and the other going beyond. Since we cannot be pixel-perfect, the default lines should be a bit wider than the bar.

At a closer look, you are right that .15, .85 work for the vertical slider but not very well for the horizontal slider. I'd be ok if you tune the horizontal slider by using .2, .9 and adding a comment on that. Interestingly, this asymmetric axes coordinates result in a symmetric extension of the line. It seems that for some reason, horizontal and vertical coordinate to pixel mapping is different. If somebody is interested, you're welcome to dig in this. But just using the empirical values is good enough for now. In the worst case we accidentially fix the coordinates and the line will move one pixel again, but that would also not be the end of the world.

**Comment by ianhi** [comment]
>  Actually, I was waiting on a code change here. 

ack! sorry for misinterpreting. 

> At a closer look, you are right that .15, .85 work for the vertical slider but not very well for the horizontal slider. I'd be ok if you tune the horizontal slider by using .2, .9 and adding a comment on that. Interestingly, this asymmetric axes coordinates result in a symmetric extension of the line. It seems that for some reason, horizontal and vertical coordinate to pixel mapping is different. If somebody is interested, you're welcome to dig in this. 

Oh that's wild. Thanks for finding those values! I've gone with the simple solution for now.

**Comment by timhoffm** [comment]
This should still get a what's new entry (see doc/users/next_whats_new/README.rst)

**Comment by timhoffm** [comment]
Milestoning as this is ready modulo the what's new entry, and we should not miss the next minor release.

**Comment by ianhi** [comment]
@timhoffm sorry I stalled out here. I was asking on gitter awhile ago about how to show the style in the whats new and @tacaswell expressed concern that these changes are somewhat backwards incompatible as there is not easy way to get the old style back.

**Comment by timhoffm** [comment]
Hm ok, maybe that's to be discussed on the dev call. OTOH, this is essentially UI element style change, which I consider less invasive than a plot style change (Any OS I'm using is changing its UI under me all the time).

**Comment by timhoffm** [comment]
Discussed in the dev call today: Widgets apperance is not considered stable API. We can change this in minor releases.

**Comment by ianhi** [comment]
Rebased to pick up fix for failing tests #20488 

Rendered What's new Entry here: https://61099-1385122-gh.circle-artifacts.com/0/doc/build/html/users/next_whats_new/slider_styling.html

**Comment by jklymak** [comment]
Any other widget users want to bless this?  Please squash commits unless @ianhi thinks they are all meaningful.

**Comment by ianhi** [comment]
> Any other widget users want to bless this?

Maybe @snowtechblog @redeboer @t-makaro 

> Please squash commits unless @ianhi thinks they are all meaningful.

They are not - I can squash

**Review by jklymak — APPROVED** [review]
Not an expert on widgets, but I think @timhoffm gave this a through review.

**Comment by BSNayal** [comment]
@ianhi @timhoffm , This would be taken as a separate bug ?
Slider_initial_position_error
